lib/graphql/static_validation/rules/fields_will_merge.rb

```diff
@@ -239,13 +232,50 @@ def find_conflict(response_key, field1, field2, mutually_exclusive: false)
           end
         end
 
+        if !conflicts[:field].key?(response_key) &&
+            (t1 = field1.definition&.type) &&
+            (t2 = field2.definition&.type) &&
+            return_types_conflict?(t1, t2)
+
+          conflict = conflicts[:return_type][response_key]
+          conflict.add_conflict(node1, "`#{t1.to_type_signature}`")
+          conflict.add_conflict(node2, "`#{t2.to_type_signature}`")
+
+          @conflict_count += 1
+        end
+
         find_conflicts_between_sub_selection_sets(
           field1,
           field2,
           mutually_exclusive: are_mutually_exclusive,
         )
       end
 
+      def return_types_conflict?(type1, type2)
+        if type1.list?
+          if type2.list?
+            return_types_conflict?(type1.of_type, type2.of_type)
+          else
+            true
+          end
+        elsif type2.list?
+        elsif type1.non_null?
+          if type2.non_null?
+            return_types_conflict?(type1.of_type, type2.of_type)
+          else
+            true
+          end
+        elsif type2.non_null?
+          true
+        elsif !type1.kind.fields? && !type2.kind.fields?
```

Comment:
Again comparing this with `graphql-js`, I believe this should be an OR...

```js
if (isLeafType(type1) || isLeafType(type2)) {
    return type1 !== type2;
  }
```

So `!type1.kind.fields?` should be equivalent to `isLeafType(type1)`, at which time the compound condition shouldn't need to change.

Reply:
Revisiting this again today, I've noticed another issue with the `fields?` check in general... Right now, [Union does not belong to `fields`](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/type_kinds.rb#L70) which makes the `fields?` check a rather ambiguous slice across both Leaf and Composite distinctions. Technically Union should probably belong to `fields?`, although I imagine that change could have a huge blast radius across community implementations.

My suggestion then would be to introduce a proper `leaf?` distinction (ie: `scalar? || enum?`), and then refactor the core library to revolve around formal `leaf?` and `composite?` distinctions, and allow `fields?` to fade away into the sunset (while remaining present to support existing community usage).

Reply:
https://github.com/rmosolgo/graphql-ruby/issues/4351